crates/pessimistic-proof/src/local_state.rs

```diff
@@ -16,38 +17,126 @@ pub struct LocalNetworkState {
+    pub fn apply_batch_header(
         &mut self,
-        certificate: &Certificate,
-    ) -> Result<(ExitRoot, BalanceRoot), ProofError> {
+        batch_header: &BatchHeader<Keccak256Hasher>,
+    ) -> Result<(), ProofError> {
         // Check the initial state
         let computed_root = self.exit_tree.get_root();
-        if computed_root != certificate.prev_local_exit_root {
-            return Err(ProofError::InvalidLocalExitRoot {
+        if computed_root != batch_header.prev_local_exit_root {
+            return Err(ProofError::InvalidInitialLocalExitRoot {
                 got: computed_root,
-                expected: certificate.prev_local_exit_root,
+                expected: batch_header.prev_local_exit_root,
             });
         }
+        if self.balance_tree.root != batch_header.prev_balance_root {
+            return Err(ProofError::InvalidInitialBalanceRoot);
+        }
+        if self.nullifier_set.root != batch_header.prev_nullifier_root {
+            return Err(ProofError::InvalidInitialNullifierRoot);
+        }
+
+        // TODO: check batch_header.imported_exits_root
+        // Apply the imported bridge exits
+        for (imported_bridge_exit, local_balance_path, nullifier_path) in
+            &batch_header.imported_bridge_exits
+        {
+            if matches!(imported_bridge_exit.bridge_exit.leaf_type, LeafType::Message) {
+                // TODO: handle bridge messages
+                panic!()
+            }
+
+            if imported_bridge_exit.sending_network == batch_header.origin_network {
+                // We don't allow a chain to exit to itself
+                return Err(ProofError::ExitToSameNetwork);
+            }
+            // Check the LER
+            if batch_header.imported_local_exit_roots[&imported_bridge_exit.sending_network]
+                != imported_bridge_exit.imported_local_exit_root
+            {
+                return Err(ProofError::InvalidImportedBridgeExitRoot);
+            }
+            // Check the LER inclusion path
+            if !imported_bridge_exit.verify_path() {
+                return Err(ProofError::InvalidImportedBridgeExitMerklePath);
+            }
+
+            // Check the nullifier non-inclusion path and update the nullifier set
+            let nullifier_key = NullifierKey {
+                network_id: imported_bridge_exit.sending_network,
+                let_index: imported_bridge_exit.leaf_index,
+            };
+            self.nullifier_set.verify_and_update(nullifier_key, nullifier_path)?;
+
+            if batch_header.origin_network
+                == imported_bridge_exit.bridge_exit.token_info.origin_network
+            {
+                // When the token is native to the chain, we don't care about the local balance
+                continue;
+            }
+            // Check the balance tree inclusion path and update the balance tree
+            let old_balance = imported_bridge_exit.current_balance;
```

Comment:
Totally valid to do it this way, but I wonder if in the future it's simpler to omit current_balance from ImportedBridgeExit, sum over BridgeExit and ImportedBridgeExits to get net flows for each token address, and then load the current balances for each affected token address.

Would save the AL service from needing to calculate and insert current balances for each ImportedBridgeExit, though this isn't a big deal.

Reply:
Yes good idea! Would save a lot of hashing when a lot of events are on the same token. Will implement it 👍

Reply:
Implemented in 892714b. It reduced the number of SP1 cycles by 50% in the simple E2E test.
It still isn't totally optimal, as I verify and update each token Merkle path individually instead of using a batch method. Added a TODO for that and will tackle it in a future PR.
Also, using the full 192-bit token info (160 for the address + 32 for the network id) is probably not optimal too. Using some sort of indexed Merkle tree would probably be more efficient. Same for the nullifier tree. Will also look into that in a future PR.

Reply:
Niiiiice!